File: sig-py/src/sig/core/fs.py
from __future__ import annotations

from pathlib import Path
from typing import NamedTuple, Protocol

SIG_DIR = ".sig"


class DirEntry(NamedTuple):
    name: str
    is_dir: bool


class SigFS(Protocol):
    def read_file(self, path: str) -> str: ...

    def write_file(self, path: str, content: str) -> None: ...

    def append_file(self, path: str, content: str) -> None: ...

    def mkdir(self, path: str, *, parents: bool = False) -> None: ...

    def readdir(self, path: str) -> list[DirEntry]: ...

    def unlink(self, path: str) -> None: ...

    def exists(self, path: str) -> bool: ...


class SigContext(NamedTuple):
    sig_dir: str
    fs: SigFS


class PathLibFS:
    def read_file(self, path: str) -> str:
        return Path(path).read_text("utf-8")

    def write_file(self, path: str, content: str) -> None:
        Path(path).write_text(content, encoding="utf-8")

    def append_file(self, path: str, content: str) -> None:
        with open(path, "a", encoding="utf-8") as file:
            file.write(content)

    def mkdir(self, path: str, *, parents: bool = False) -> None:
        Path(path).mkdir(parents=parents, exist_ok=parents)

    def readdir(self, path: str) -> list[DirEntry]:
        return [
            DirEntry(name=entry.name, is_dir=entry.is_dir())
            for entry in Path(path).iterdir()
        ]

    def unlink(self, path: str) -> None:
        Path(path).unlink()

    def exists(self, path: str) -> bool:
        return Path(path).exists()


def create_sig_context(
    project_root: str,
    *,
    fs: SigFS | None = None,
    sig_dir: str | None = None,
) -> SigContext:
    return SigContext(
        sig_dir=sig_dir if sig_dir is not None else str(Path(project_root) / SIG_DIR),
        fs=fs or PathLibFS(),
    )


def to_sig_context(ctx_or_sig_dir: SigContext | str) -> SigContext:
    if isinstance(ctx_or_sig_dir, str):
        return SigContext(sig_dir=ctx_or_sig_dir, fs=PathLibFS())
    return ctx_or_sig_dir

File: sig-ts/src/core/fs.ts
import {
  access as fsAccess,
  appendFile as fsAppendFile,
  mkdir as fsMkdir,
  readdir as fsReaddir,
  readFile as fsReadFile,
  unlink as fsUnlink,
  writeFile as fsWriteFile,
} from 'node:fs/promises';
import { join } from 'node:path';
import type { Dirent, SigContext, SigFS } from '../types.js';

const SIG_DIR = '.sig';

export class NodeFS implements SigFS {
  async readFile(path: string, encoding: 'utf8'): Promise<string> {
    return fsReadFile(path, encoding);
  }

  async writeFile(path: string, content: string, encoding: 'utf8'): Promise<void> {
    await fsWriteFile(path, content, encoding);
  }

  async appendFile(path: string, content: string, encoding: 'utf8'): Promise<void> {
    await fsAppendFile(path, content, encoding);
  }

  async mkdir(path: string, options?: { recursive?: boolean }): Promise<void> {
    await fsMkdir(path, options);
  }

  async readdir(path: string, options: { withFileTypes: true }): Promise<Dirent[]> {
    return fsReaddir(path, options);
  }

  async unlink(path: string): Promise<void> {
    await fsUnlink(path);
  }

  async access(path: string): Promise<void> {
    await fsAccess(path);
  }
}

export function createSigContext(
  projectRoot: string,
  options?: { fs?: SigFS; sigDir?: string }
): SigContext {
  return {
    sigDir: options?.sigDir ?? join(projectRoot, SIG_DIR),
    fs: options?.fs ?? new NodeFS(),
  };
}

export function toSigContext(ctxOrSigDir: SigContext | string): SigContext {
  if (typeof ctxOrSigDir === 'string') {
    return {
      sigDir: ctxOrSigDir,
      fs: new NodeFS(),
    };
  }
  return ctxOrSigDir;
}
